Add support for hierarchical list prompts

patriksvensson opened the pull request:
This PR introduces support for hierarchical list prompts.

There is one breaking change here. Items in multi-selection prompts can no longer be configured to be pre-selected using an index since this doesn't make sense anymore.

![prompt_tree](https://user-images.githubusercontent.com/357872/118544572-d8109280-b755-11eb-8848-acf48a92a19d.gif)

Closes #412

phil-scott-78 reviewed (commented):
Code looks good, but I'm going back and forth with the `Hierarchical` property name. Without actually running the code it wasn't terribly clear what would happen between setting it vs `NonHierarchical`. Hierarchical to me indicates that there is going to be a ranking, but nothing that the parents being selectedable.

I kind of like how Vuetify names these two modes "leaf" and "independent" where leaf mode you can only select the leafs vs independent which allows parents to be selected too.

patriksvensson commented:
@phil-scott-78 That's some good feedback. I've added a `SelectionMode` enum that contains `Leaf` and `Independent` like Vuetify has, which I think makes sense. I've also added selection mode support for `SelectionPrompt` as well.

phil-scott-78 commented on examples/Console/Prompt/Program.cs:
```diff
@@ -11,6 +13,26 @@ public static void Main(string[] args)
                 return;
             }
 
+            var favorites = AnsiConsole.Prompt(
```
I missed this when doing the first pass on the review. I suspect we didn't want this added.

patriksvensson replied:
You didn't miss it. I accidentally left it in the latest update 😅

patriksvensson replied:
@phil-scott-78 Removed!
